pkg/sql/contention/txnidcache/writer_test.go

```diff
@@ -0,0 +1,131 @@
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package txnidcache
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"sync"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+)
+
+type blackHoleSink struct {
+	pool *sync.Pool
+
+	// Simulate a real sink
+	ch chan *messageBlock
+}
+
+var _ messageSink = &blackHoleSink{}
+
+func newBlackHoleSink(pool *sync.Pool, chanSize int) *blackHoleSink {
+	return &blackHoleSink{
+		pool: pool,
+		ch:   make(chan *messageBlock, chanSize),
+	}
+}
+
+func (b *blackHoleSink) start() {
+	go func() {
+		for block := range b.ch {
+			*block = messageBlock{}
+			b.pool.Put(block)
+		}
+	}()
+}
+
+func (b *blackHoleSink) stop() {
+	close(b.ch)
+}
+
+// push implements messageSink interface.
+func (b *blackHoleSink) push(block *messageBlock) {
+	b.ch <- block
+}
+
+// generateUUID uses a provided integer to populate uuid.UUID. This is
+// to avoid UUID generation slowing down the benchmark.
+func generateUUID(i uint64) uuid.UUID {
+	id := uuid.UUID{}
+	binary.LittleEndian.PutUint64(id[:], i)
+	return id
+}
+
+func BenchmarkWriter(b *testing.B) {
+	skip.UnderShort(b)
+	defer log.Scope(b).Close(b)
+
+	ctx := context.Background()
+
+	for _, sinkType := range []string{"blackHoleSink", "realSink"} {
+		b.Run(fmt.Sprintf("sinkType=%s", sinkType), func(b *testing.B) {
+			for _, numOfConcurrentWriter := range []int{1, 24, 48, 64, 92, 128} {
+				b.Run(fmt.Sprintf("concurrentWriter=%d", numOfConcurrentWriter), func(b *testing.B) {
+					starter := make(chan struct{})
+
+					blockPool := &sync.Pool{
+						New: func() interface{} {
+							return &messageBlock{}
+						},
+					}
+					var sink messageSink
+
+					switch sinkType {
+					case "blackHoleSink":
+						blackHole := newBlackHoleSink(blockPool, channelSize)
+						blackHole.start()
+						defer blackHole.stop()
+						sink = blackHole
+					case "realSink":
+						st := cluster.MakeTestingClusterSettings()
+						metrics := NewMetrics()
+						realSink := NewTxnIDCache(st, &metrics)
+						realSink.Start(ctx, stop.NewStopper())
+						sink = realSink
+					default:
+						panic(fmt.Sprintf("invalid sink type: %s", sinkType))
+					}
+
+					w := newWriter(sink, blockPool)
+
+					b.ResetTimer()
+					b.SetBytes(messageBlockSize * entrySize)
+					var wg sync.WaitGroup
+					for writerIdx := 0; writerIdx < numOfConcurrentWriter; writerIdx++ {
+						wg.Add(1)
+
+						go func() {
+							defer wg.Done()
+
+							<-starter
+
+							for i := 0; i < b.N; i++ {
```

Comment:
it's probably interesting to divide `b.N` by `numOfConcurrentWriter`

Reply:
Done. 

The updated benchmark number

```
$ benchstat ~/tmp/blackhole ~/tmp/real                                                                                                                           [14:58:33]
name                            old time/op    new time/op    delta
Writer/concurrentWriter=1-24      72.2ns ± 0%   788.1ns ± 6%  +991.61%  (p=0.000 n=9+10)
Writer/concurrentWriter=24-24      117ns ±11%     227ns ± 9%   +94.19%  (p=0.000 n=10+10)
Writer/concurrentWriter=48-24      123ns ±16%     176ns ± 9%   +43.10%  (p=0.000 n=10+9)
Writer/concurrentWriter=64-24      122ns ± 2%     181ns ± 2%   +47.90%  (p=0.000 n=8+8)
Writer/concurrentWriter=92-24      142ns ± 8%     164ns ± 4%   +15.06%  (p=0.000 n=10+10)
Writer/concurrentWriter=128-24     160ns ± 7%     174ns ± 5%    +8.74%  (p=0.019 n=10+10)

name                            old speed      new speed      delta
Writer/concurrentWriter=1-24     340GB/s ± 0%    31GB/s ± 6%   -90.83%  (p=0.000 n=9+10)
Writer/concurrentWriter=24-24    211GB/s ±10%   108GB/s ± 9%   -48.61%  (p=0.000 n=10+10)
Writer/concurrentWriter=48-24    201GB/s ±14%   140GB/s ± 9%   -30.47%  (p=0.000 n=10+9)
Writer/concurrentWriter=64-24    201GB/s ± 2%   136GB/s ± 2%   -32.39%  (p=0.000 n=8+8)
Writer/concurrentWriter=92-24    173GB/s ± 8%   150GB/s ± 4%   -13.38%  (p=0.000 n=10+10)
Writer/concurrentWriter=128-24   154GB/s ± 7%   142GB/s ± 5%    -8.15%  (p=0.019 n=10+10)

name                            old alloc/op   new alloc/op   delta
Writer/concurrentWriter=1-24       0.00B        219.00B ± 3%     +Inf%  (p=0.000 n=10+9)
Writer/concurrentWriter=24-24      0.00B         22.60B ± 3%     +Inf%  (p=0.000 n=10+10)
Writer/concurrentWriter=48-24      0.00B         19.00B ± 0%     +Inf%  (p=0.000 n=10+10)
Writer/concurrentWriter=64-24      0.00B         19.00B ± 0%     +Inf%  (p=0.000 n=10+10)
Writer/concurrentWriter=92-24      0.00B         17.60B ± 3%     +Inf%  (p=0.000 n=10+10)
Writer/concurrentWriter=128-24     0.00B         17.30B ± 4%     +Inf%  (p=0.000 n=10+10)

name                            old allocs/op  new allocs/op  delta
Writer/concurrentWriter=1-24        0.00           1.00 ± 0%     +Inf%  (p=0.000 n=10+10)
Writer/concurrentWriter=24-24       0.00           1.00 ± 0%     +Inf%  (p=0.000 n=10+10)
Writer/concurrentWriter=48-24       0.00           1.00 ± 0%     +Inf%  (p=0.000 n=10+10)
Writer/concurrentWriter=64-24       0.00           0.00           ~     (all equal)
Writer/concurrentWriter=92-24       0.00           0.00           ~     (all equal)
Writer/concurrentWriter=128-24      0.00           0.00           ~     (all equal)
```

Reply:
CPU Profile 

<img width="2367" alt="Screen Shot 2022-02-08 at 17 59 42" src="https://user-images.githubusercontent.com/9267198/153089990-1faa9bd8-63cc-46dd-9e91-f8da94dee360.png">

Memory Profile

<img width="2367" alt="Screen Shot 2022-02-08 at 17 55 44" src="https://user-images.githubusercontent.com/9267198/153089997-f93e1f01-0e10-4f72-922b-323d2eb2b212.png">

Heap Profile

<img width="2367" alt="Screen Shot 2022-02-08 at 18 03 07" src="https://user-images.githubusercontent.com/9267198/153090344-0cd4b38b-c53f-4d5e-93e0-4bd24ca2e3d1.png">



I'm not too sure how to interpret the graphs here. It seems like there is a lot of time spent in the `cache.UnorderedCache` to perform eviction, and there some allocations performed in `cache.UnorderedCache`. 

I'm not sure if all the small allocation made by the `cache.UnorderedCache` can cause more GC pressure? My reasoning here is that the `kv` workload is throwing nearing 20k QPS and since we perform 2 writes per operation, we are performing 40k writes per second. That would be a lot of small object allocated on the heap (used for the linked list used to perform FIFO eviction).

Reply:
I did another kv95 run by turning `txnidcache.Cache` into effectively a no-op sink. It actually performs on par with before TxnIDCache was introduced.

| Branch                | elapsed       | errors | ops(total)  | ops/sec(cum) | avg(ms) | p50(ms) | p95(ms)  | p99(ms)  | pMax(ms)  |
|-----------------------|---------------|--------|-------------|--------------|---------|---------|----------|----------|-----------|
| baseline              | 2700.0s       |   0    | 55785736    |  20661.4     | 3.1     |  2.4    |  8.4     |  13.1    |  75.5     |
| cache-normal         | 2700.0s       |   0    | 49867068    |  18469.3     | 3.5     |  2.5    |  10.0    |  17.8    |  671.1    |
| cache-noop-sink       | 2700.0s       |   0    | 56440227    |  20903.8     | 3.1     |  2.4    |  8.1     |  13.1    |  71.3     |


I think at this point it's reasonable to believe that the slowdown comes from the sink which couldn't keep up with the incoming writes, causing the channel to fill up which resulted in back pressure and cause the `concurrentWriteBuffer` to wait.

I think at this point there are a few options:
1. have multiple `cache.UnorderedCache` and each cache has its own goroutine / channel.
2. try out using the ring buffer. This might introduce slightly slower read, but I think since read doesn't happen during the query execution hot path, it's an acceptable trade off.
3. maybe both?

I will try out these two ideas tomorrow.

Reply:
Here was the diff I introduced

```
diff --git a/pkg/sql/contention/txnidcache/txn_id_cache.go b/pkg/sql/contention/txnidcache/txn_id_cache.go
index 9384ed7044..e335ec6e39 100644
--- a/pkg/sql/contention/txnidcache/txn_id_cache.go
+++ b/pkg/sql/contention/txnidcache/txn_id_cache.go
@@ -171,14 +171,15 @@ func NewTxnIDCache(st *cluster.Settings, metrics *Metrics) *Cache {
 // Start implements the Provider interface.
 func (t *Cache) Start(ctx context.Context, stopper *stop.Stopper) {
 	addBlockToStore := func(msgBlock *messageBlock) {
-		t.mu.Lock()
-		defer t.mu.Unlock()
-		for blockIdx := range msgBlock {
-			if !msgBlock[blockIdx].valid() {
-				break
-			}
-			t.mu.store.Add(msgBlock[blockIdx].TxnID, msgBlock[blockIdx].TxnFingerprintID)
-		}
+		// TODO(azhng): wip: imagine if we are using a blackhoel sink
+		// t.mu.Lock()
+		// defer t.mu.Unlock()
+		// for blockIdx := range msgBlock {
+		// 	if !msgBlock[blockIdx].valid() {
+		// 		break
+		// 	}
+		// 	t.mu.store.Add(msgBlock[blockIdx].TxnID, msgBlock[blockIdx].TxnFingerprintID)
+		// }
 	}
 
 	consumeBlock := func(b *messageBlock) {
```